### PR TITLE
Fixing issue #5482: Port from desktop .NET: Performance improvement for XmlUTF8TextReader when reading invalid chars

### PR DESCRIPTION
**Pull request by SGuyGe**
Perf comparison    Before                               After
TestEFBytes             ~360 ms, 10 iterations   ~35 ms, 29 iterations
TestAsciiBytes        ~7.x ms, 137 iterations   ~7.x ms, 123 iteration

@shmao @khdang @zhenlan

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextReader.cs [hunk 1]
```diff
@@ -892,6 +892,24 @@ private void ReadAttributes()
             ProcessAttributes();
         }
 
+        // NOTE: Call only if 0xEF has been seen in the stream AND there are three valid bytes to check (buffer[offset], buffer[offset + 1], buffer[offset + 2]). 
+        // 0xFFFE and 0xFFFF are not valid characters per Unicode specification. The first byte in the UTF8 representation is 0xEF. 
+        bool IsNextCharacterNonFFFE(byte[] buffer, int offset)
```
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextReader.cs]: Afaik, as convention we specify method scope keywords explicitly for method so we should use ```private``` here. Similar comments for the newly added methods.

File: src/System.Runtime.Serialization.Xml/tests/Performance/XmlUTF8TextReaderTests.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+using System.Xml;
+
+namespace System.Runtime.Serialization.Xml.Tests.Performance
+{
+    public class XmlUTF8TextReaderTests
+    {
+        // <Root>Testing</Root>
+        static byte[] trivial = new byte[] { 0x3C, 0x52, 0x6F, 0x6F, 0x74, 0x3E, 0x54, 0x65, 0x73, 0x74, 0x69, 0x6E, 0x67, 0x3C, 0x2F, 0x52, 0x6F, 0x6F, 0x74, 0x3E };
+
+        // <Root>?????</Root>
```
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/XmlUTF8TextReaderTests.cs]: Nit: Does the ? just to indicate a character or it doesn't show up correctly?

File: src/System.Runtime.Serialization.Xml/tests/Performance/XmlUTF8TextReaderTests.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Xml;
+
+namespace System.Runtime.Serialization.Xml.Tests.Performance
+{
+    public class XmlUTF8TextReaderTests
+    {
+        // <Root>Testing</Root>
+        static byte[] trivial = new byte[] { 0x3C, 0x52, 0x6F, 0x6F, 0x74, 0x3E, 0x54, 0x65, 0x73, 0x74, 0x69, 0x6E, 0x67, 0x3C, 0x2F, 0x52, 0x6F, 0x6F, 0x74, 0x3E };
```
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/XmlUTF8TextReaderTests.cs]: Nit: I think we can use ```private``` for the ```static```s as well

**Comment by khdang** [comment]
LGTM. Somehow I don't see any checks running for this PR, maybe it's only me.

File: src/System.Runtime.Serialization.Xml/tests/Performance/XmlUTF8TextReaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            byteArray[offset++] = 0x6F;     // o
+            byteArray[offset++] = 0x74;     // t
+            byteArray[offset++] = 0x3E;     // >
+
+            return byteArray;
+        }
+
+        const int FiveMega = 5 * 1024 * 1024;
+
+        public void RunTest(byte[] testBytes, int iterations)
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/Performance/XmlUTF8TextReaderTests.cs]: The method name 'RunTest' seems unspecific, considering it's in the class named XmlUTF8TextReaderTests. Can it be 'RunPerfTest' or something more meaningful?

**Comment by shmao** [comment]
LGTM.

**Pull request merged**